### DISCUSSION TITLE
simplest things seem fraught with mysterious errors

### DISCUSSION THREAD

**Discussion by lewisl** [xmake-io/xmake#3616] [General]
I am just trying to learn and I am thwarted if I go a step past the velvet ropes. Here is the error:

```
Executing task: xmake -w 

[ 25%]: cache compiling.debug src/main.cpp
error: error: unknown target triple 'unknown-apple-macosx13.3.0', please use -triple or -arch
  > in src/main.cpp
```

Here is xmake.lua:
```
add_rules("mode.debug", "mode.release")

target("composition")
    set_kind("binary")
    add_files("src/*.cpp")
    add_includedirs("include")
    set_languages("cxx14")
    set_optimize("faster")
```

Directory structure:
```
~/.../composition
           /.cache    #  stuff clangd creates
          /.xmake     # project config
          /build
          /include
                 composition.h
          /src
                 main.cpp
         .gitignore
         xmake.lua

**Comment by lewisl** [comment]
Here is some additional output with the verbose option:

```
composition xmake b -v
[ 25%]: cache compiling.debug src/main.cpp
/Library/Developer/CommandLineTools/usr/bin/clang -c -Qunused-arguments -target undefined-apple-macos13.3 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk -std=c++14 -Iinclude -o build/.objs/composition/macosx/undefined/debug/src/main.cpp.o src/main.cpp
```
Look at the -target option.  Where the heck did that come from? It's not part of the project description that I created.  This is some spurious thing xmake generated.  Why?   How to correct it?   Seems like I should report as bug.  I have 2 other projects working fine without this problem.

**Comment by lewisl** [comment]
Here is the problem:  xmake built compile_commands.json in a completely bogus way.  And this results in the errors when running clang.

```
[
{
  "directory": "/Users/lewis/Dropbox/Online Coursework/c++ experiments/composition",
  "arguments": ["/Library/Developer/CommandLineTools/usr/bin/clang", "-c", "-Qunused-arguments", "-target", "undefined-apple-macos13.3", "-isysroot", "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk", "-std=c++14", "-Iinclude", "-o", "build/.objs/composition/macosx/undefined/debug/src/main.cpp.o", "src/main.cpp"],
  "file": "src/main.cpp"
}]
```

So, why did it do this?   Something caused the wrong arguments to be created. I have never manually edited one of these.  Indeed, that would be a mistake because xmake will write over it each time xmake.lua is edited.

**Comment by lewisl** [comment]
So, I manually fixed compile_commands.json by replacing ```"undefined-apple-macos13.3"```     with```"arm64-apple-macos13.3"```.  But, silly me:  this does nothing.  xmake doesn't use this json file to drive the build.  This is just a dump of the parameters for other applications to read.

**Comment by lewisl** [comment]
So, I went through the curses interface for configuration.  This is the first time ever.  For some impenetrable reason the target for this project had been set for watchOS.  I manually changed it to macos and everything works.

I would NOT consider this closed.  How did this ever happen?  How did the auto-recognition of the build platform and the target fail?  Let's get to the bottom of this.

**Comment by lewisl** [comment]
It seems to me that the vs code extension is simply massively broken.  Sometimes it works and sometimes it is a NOOP.  I just uninstalled and re-installed.  I added "add_rules("mode.debug", "mode.release")" to the top of xmake.lua.  I used the extension to build.  No release build was created.

Basically, only the command line works.

I want to love xmake.  It's the best designed build tool. But, it is flaky.

**Reply by waruqi** [reply]
it's xmake-vscode plugin issue, it does not support macos arm64 yet.

https://github.com/xmake-io/xmake-vscode/blob/734c0741062ae519cfcd233b1446ef85dff38f15/src/xmake.ts#L1159


default x86_64

you need select arch to arm64 in bottom toolbar

**Comment by lewisl** [comment]
OK.  Sorry for all the notes.  Took me a while to see what the problem was.

Command line is more reliable.  But, why did it work for others of my projects with no problem?

From: ruki ***@***.***>
Reply-To: xmake-io/xmake ***@***.***>
Date: Monday, April 10, 2023 at 3:40 PM
To: xmake-io/xmake ***@***.***>
Cc: Lewis Levin ***@***.***>, Author ***@***.***>
Subject: Re: [xmake-io/xmake] simplest things seem fraught with mysterious errors (Discussion #3616)


it's xmake-vscode plugin issue, it does not support macos arm64 yet.

**Comment by waruqi** [comment]
I have improve vscode plugin to detect arm64 arch

**Comment by lewisl** [comment]
Thanks.  Will give it a try.

I think I built my other projects with the command line tool and then used vscode extension to build/run while working on the code.

From: ruki ***@***.***>
Reply-To: xmake-io/xmake ***@***.***>
Date: Monday, April 10, 2023 at 6:57 PM
To: xmake-io/xmake ***@***.***>
Cc: Lewis Levin ***@***.***>, Author ***@***.***>
Subject: Re: [xmake-io/xmake] simplest things seem fraught with mysterious errors (Discussion #3616)


I have improve vscode plugin to detect arm64 arch
